src/Mvc/Mvc.Analyzers/src/TopLevelParameterNameAnalyzer.cs

```diff
@@ -185,20 +183,100 @@ internal static bool SpecifiesModelType(in SymbolCache symbolCache, IParameterSy
 
         internal readonly struct SymbolCache
         {
-            public SymbolCache(Compilation compilation)
+            public SymbolCache(
+                INamedTypeSymbol bindAttribute,
+                INamedTypeSymbol controllerAttribute,
+                INamedTypeSymbol fromBodyAttribute,
+                INamedTypeSymbol apiBehaviorMetadata,
+                INamedTypeSymbol binderTypeProviderMetadata,
+                INamedTypeSymbol modelNameProvider,
+                INamedTypeSymbol nonControllerAttribute,
+                INamedTypeSymbol nonActionAttribute,
+                IMethodSymbol disposableDispose)
             {
-                BindAttribute = compilation.GetTypeByMetadataName(SymbolNames.BindAttribute);
-                ControllerAttribute = compilation.GetTypeByMetadataName(SymbolNames.ControllerAttribute);
-                FromBodyAttribute = compilation.GetTypeByMetadataName(SymbolNames.FromBodyAttribute);
-                IApiBehaviorMetadata = compilation.GetTypeByMetadataName(SymbolNames.IApiBehaviorMetadata);
-                IBinderTypeProviderMetadata = compilation.GetTypeByMetadataName(SymbolNames.IBinderTypeProviderMetadata);
-                IModelNameProvider = compilation.GetTypeByMetadataName(SymbolNames.IModelNameProvider);
-                NonControllerAttribute = compilation.GetTypeByMetadataName(SymbolNames.NonControllerAttribute);
-                NonActionAttribute = compilation.GetTypeByMetadataName(SymbolNames.NonActionAttribute);
+                BindAttribute = bindAttribute;
+                ControllerAttribute = controllerAttribute;
+                FromBodyAttribute = fromBodyAttribute;
+                IApiBehaviorMetadata = apiBehaviorMetadata;
+                IBinderTypeProviderMetadata = binderTypeProviderMetadata;
+                IModelNameProvider = modelNameProvider;
+                NonControllerAttribute = nonControllerAttribute;
+                NonActionAttribute = nonActionAttribute;
+                IDisposableDispose = disposableDispose;
+            }
+
+            public static bool TryCreate(Compilation compilation, out SymbolCache symbolCache)
+            {
+                symbolCache = default;
+
+                if (!TryGetType(SymbolNames.BindAttribute, out var bindAttribute))
+                {
+                    return false;
+                }
+
+
+                if (!TryGetType(SymbolNames.ControllerAttribute, out var controllerAttribute))
+                {
+                    return false;
+                }
+
+
```

Comment:
@JamesNK

Reply:
![image](https://user-images.githubusercontent.com/303201/58130581-76091980-7c70-11e9-89b9-f1cbbde53ed8.png)

Reply:
Why is aquaman sad?

Reply:
Thanos collected the one ring and now aquaman has to leave Hogwarts. Poor aquaman 😢

Reply:
I'll clean up the extra new lines in a follow up.